File: Surya_Prakash/Day 3/Search in a rotated sorted array.py
def search(self, nums: List[int], t: int) -> int:
    n = len(nums)

    # find peak
    l = -1
    h = n
    while h - l > 1:
        mid = (l + h) // 2
        if nums[mid] > nums[-1]:
            l = mid
        else:
            h = mid
    peak = h
    print(nums[peak], nums[l])

    # setting ends for searching target
    # if nums[peak]==t:return peak
    if t <= nums[-1]:
        l = peak
        h = n - 1
    else:
        l = 0
        h = peak - 1
    '''
    while l<=h:
        mid=(l+h)//2
        if nums[mid]==t:return mid
        elif t>nums[mid]:
            l=mid+1
        else:
            h=mid-1
    return -1
    '''

    l -= 1
    while h - l > 1:
        mid = (l + h) // 2
        if t > nums[mid]:
            l = mid
        else:
            h = mid
    return h if nums[h] == t else -1

File: parasmittal/Day 3/Search in a rotated sorted array.cpp
class Solution {
public:
    int util(vector<int> arr,int l,int r,int target){
        if(l>r) return -1;
        if(l==r && arr[l]==target) return l;
        if(l==r) return -1;
        int mid=r-(r-l)/2;
        if(arr[mid]==target) return mid;
        if(arr[mid]<arr[l]){
            if(target<=arr[r] && target>=arr[mid])
                return util(arr,mid+1,r,target);
        return util(arr,l,mid-1,target);
            
        }
        if(target>=arr[l] && target <=arr[mid])
        return util(arr,l,mid-1,target);
        return util(arr,mid+1,r,target);
    }
    int search(vector<int>& nums, int target) {
        int n=nums.size();
        return util(nums,0,n-1,target);
    }
};